### PR TITLE
Handle changes token in the querystring.

### PR DESCRIPTION
**Pull request by frankban**
Import a collection of bundle changes from
a token reference identifying those changes.

### CODE REVIEW

**Comment by frankban** [comment]
QA:
- set up an environment including the GUI charm in development:
  `juju-quickstart --gui-charm-url cs:~juju-gui/trusty/juju-gui-49`;
- wait for the environment to be ready;
- switch to this branch: 
  `juju set juju-gui juju-gui-source="https://github.com/frankban/juju-gui.git changeset-token"`;
- wait for the new build to be ready;
- ensure the GUI works as expected, no JS errors, etc...
- to live test the uncommitted bundle deployment, let's use a work in progress
  quickstart prototype: 
  `bzr branch lp:~frankban/juju-quickstart/uncommitted-bundles quickstart` then
  `cd quickstart` and `make`;
- at this point, you can use `devenv/bin/quickstart -u` (or `--uncommitted`) to
  deploy bundles opening the GUI with the uncommitted state, e.g.: 
  `devenv/bin/juju-quickstart openstack-telemetry -u` and
  `devenv/bin/juju-quickstart -u mediawiki-single`.
- the process above should open the GUI and after a bit all the uncommitted
  services should be displayed.

File: test/test_bundle_importer.js [hunk 1]
```diff
@@ -63,16 +63,35 @@ describe('Bundle Importer', function() {
...
+
     describe('importBundleYAML', function() {
       it('calls fetchDryRun with yaml', function() {
         var fetch = utils.makeStubMethod(bundleImporter, 'fetchDryRun');
         this._cleanups.push(fetch.reset);
         bundleImporter.importBundleYAML('foo: bar');
         assert.equal(fetch.callCount(), 1);
-        assert.equal(fetch.lastArguments()[0], 'foo: bar');
+        var args = fetch.lastArguments();
+        assert.lengthOf(args, 2);
```
Comment (hatched) [inline test/test_bundle_importer.js]: can you convert these 'fancy' assert methods to just use 'assert.equal'. Unfortunately the fancy ones give 0 traceback as they get stuck in Chai land.

**Comment by hatched** [comment]
:+1: With the assertions changed over.
No QA

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2523/
Test PASSed.

File: test/test_bundle_importer.js (hunk 1 above)
Comment (frankban) [inline test/test_bundle_importer.js]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2526/
Test PASSed.

File: app/app.js [hunk 2]
```diff
@@ -1270,6 +1270,23 @@ YUI.add('juju-gui', function(Y) {
...
         }
+        // Handle the change set token if provided in the query.
+        // The change set token identifies a collections of changes required
+        // to deploy a bundle. Those changes are assumed to be already
+        // registered in the GUI server (via a ChangeSet:SetChanges request).
+        // Doing that is usually responsibility of a separate system
+        // (most of the times, it is Juju Quickstart).
+        var querystring = this.location.search.substring(1);
+        var qs = Y.QueryString.parse(querystring);
+        var changesToken = qs.changestoken;
```
Comment (mitechie) [inline app/app.js]: what is a changestoken? is it a bundle ID? A auth token and we're expecting to get the bundle yaml from quickstart? Is it something else? This feels a little bit like using a word to define itself and I wonder if some more concrete clarity would make this easier to come back to in a year.

File: app/app.js [hunk 3]
```diff
@@ -1270,6 +1270,23 @@ YUI.add('juju-gui', function(Y) {
...
+        if (Y.Lang.isValue(changesToken)) {
+          // De-dupe if necessary.
+          if (Y.Lang.isArray(changesToken)) {
```
Comment (mitechie) [inline app/app.js]: good catch here. This is de-duping from the question string correct? Maybe a note to that effect to mark it as different than de-duping the items in the changeset itself?

File: app/store/env/go.js [hunk 4]
```diff
@@ -2236,13 +2236,15 @@ YUI.add('juju-env-go', function(Y) {
 
       @method getChangeSet
       @param {String} bundleYAML The bundle YAML file contents.
+      @param {String} changesToken The token identifying a bundle change set
```
Comment (mitechie) [inline app/store/env/go.js]: this reads that the bundleYAML is ignored if you have a changestoken, but in the if block below we first run with the yaml if it's not null?

Comment (mitechie) [inline app/store/env/go.js]: bah sorry read it backwards. This clearly states the changestoken is ignored if the yaml is provided.

Comment (mitechie) [inline app/store/env/go.js]: then I'm at a loss how we can provide the yaml file (say a local bundle) and use quickstart to pass it into the changeset and have it become uncomitted?

**Comment by mitechie** [comment]
:+1: and thanks for walking me through it and noting that this is the first time the gui gets the bundle information from quickstart.

**Comment by mitechie** [comment]
QA OK, ty for the updates and looking forward to the quckstart follow up branch.

**Comment by frankban** [comment]
Ty for your reviews and QA!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/887

**Comment by frankban** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**